### PR TITLE
Add sets.all and sets.some operators to sets

### PR DESCRIPTION
**Pull request by mudathirmahgoub**

### CODE REVIEW

File: src/theory/bags/contribution.txt [hunk 1]
```diff
@@ -0,0 +1,27 @@
+contribution
```
Comment (ajreynol) [inline src/theory/bags/contribution.txt:1]: Spurious file?

File: src/theory/sets/theory_sets_private.cpp [hunk 2]
```diff
@@ -390,6 +390,13 @@ void TheorySetsPrivate::fullEffortCheck()
     {
       continue;
     }
+    // check set.all rules
+    checkAll();
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_private.cpp:394]: This is a bit of a misleading name, can you rename this to something that doesn't imply all things are being checked rather it is for `set.all`?

**Review by ajreynol — CHANGES_REQUESTED** [review]
Generally I think this looks like a helpful operator, I've left some comments.

File: src/theory/bags/contribution.txt (hunk 1 above)
Comment (mudathirmahgoub) [reply src/theory/bags/contribution.txt:1]: Yes !

File: src/theory/sets/theory_sets_private.cpp (hunk 2 above)
Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets_private.cpp:394]: I was hoping for a suggestion for a better name. 
How about checkPredicateAll?

Comment (ajreynol) [reply src/theory/sets/theory_sets_private.cpp]: Sure

Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets_private.cpp]: already renamed.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/sets/theory_sets_private.cpp [hunk 3]
```diff
@@ -745,12 +748,13 @@ void TheorySetsPrivate::checkFilterUp()
...
 {
   NodeManager* nm = nodeManager();
+  SkolemManager* sm = nm->getSkolemManager();
   const std::vector<Node>& filterTerms = d_state.getFilterTerms();
   for (const Node& term : filterTerms)
   {
     Node p = term[0];
     Node A = term[1];
-
+    p = sm->getUnpurifiedForm(p);
```
[the remark this answers is not in the document]
Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets_private.cpp]: It seems no longer necessary with your latest PR. It is removed now

File: src/theory/sets/theory_sets.cpp [hunk 4]
```diff
@@ -81,6 +81,8 @@ void TheorySets::finishInit()
   d_equalityEngine->addFunctionKind(Kind::SET_MINUS);
   d_equalityEngine->addFunctionKind(Kind::SET_MEMBER);
   d_equalityEngine->addFunctionKind(Kind::SET_SUBSET);
+  d_equalityEngine->addFunctionKind(Kind::SET_ALL);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets.cpp:84]: You can delete these 2 lines.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/sets/theory_sets.cpp (hunk 4 above)
Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets.cpp]: Done

**Review by ajreynol — CHANGES_REQUESTED** [review]
I am wondering if we should just reduce these operators eagerly in the rewriter. This would make the reduction methods unecessary, as well as a few other things like needing to mark these as congruence kinds in the equality engine.

Do you think it is necessary to reduce these lazily?

**Comment by mudathirmahgoub** [comment]
> I am wondering if we should just reduce these operators eagerly in the rewriter. This would make the reduction methods unecessary, as well as a few other things like needing to mark these as congruence kinds in the equality engine.
> 
> Do you think it is necessary to reduce these lazily?

I think you are talking about prerewrite versus postrewrite. I don't have a solid reason and I was wondering the same. 
However, this [line](https://github.com/cvc5/cvc5/pull/10948/files#diff-e9af69b200eac4565772b194f6c68a8f91caf976e0aa572d75460569196f2d73R128) to me seems faster when the set is in normal form and one of its element falsifies the predicate, compared to checking the equality between the set and its filter. Similarly for set.some with this [line](https://github.com/cvc5/cvc5/pull/10948/files#diff-e9af69b200eac4565772b194f6c68a8f91caf976e0aa572d75460569196f2d73R138). I think this is useful for benchmarks with concrete elements. The downside is the introduction of more filter terms, which I think is fine.

**Pull request merged**